File: Problems/No.1284_Minimum_Number_of_Flips_to_Convert_Binary_Matrix_to_Zero_Matrix/minimum_number_of_flips_to_convert_binary_matrix_to_zero_matrix.py
class Solution(object):
    def minFlips(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: int
        """

        # check if it is zero matrix
        def CheckZero(matrix):
            for c_i in range(m):
                for c_j in range(n):
                    if matrix[c_i][c_j]:
                        return False
            return True
    

        # parameters
        AllPossibles = []
        checked = []
        checks = [[0, -1], [0, 1], [-1, 0], [1, 0]]
        flip_count, m, n = 0, len(mat), len(mat[0])

        # start computing
        AllPossibles.append(mat)
        checked.append(mat)
        while AllPossibles:
            CheckSize = len(AllPossibles)

            for i in range(CheckSize):
                if CheckZero(AllPossibles[0]):
                    return flip_count
                else:
                    for row in range(m):
                        for col in range(n):
                            currMat = deepcopy(AllPossibles[0])
                            currMat[row][col] = 1 - currMat[row][col]

                            for check in checks:
                                newRow = row + check[0]
                                newCol = col + check[1]
                                if ((0 <= newRow < m) and (0 <= newCol < n)):
                                    currMat[newRow][newCol] = 1 - currMat[newRow][newCol]
                            
                            if (currMat not in checked):
                                AllPossibles.append(currMat)
                                checked.append(currMat)
                AllPossibles.pop(0)
            flip_count += 1
        
        return -1

File: Problems/No.1284_Minimum_Number_of_Flips_to_Convert_Binary_Matrix_to_Zero_Matrix/minimum_number_of_flips_to_convert_binary_matrix_to_zero_matrix.cpp
class Solution {
public:

    // check if it is zero matrix
    bool CheckZero(vector<vector<int>>& matrix) {

        for (int i=0; i<matrix.size(); i++) 
            for (int j=0; j<matrix[0].size(); j++) 
                if (matrix[i][j] == 1) return false;
        
        return true;

    }

    int minFlips(vector<vector<int>>& mat) {
        
        // parameters
        std::queue<std::vector<std::vector<int>>> AllPossibles;
        std::set<std::vector<std::vector<int>>> checked; 
        std::vector<std::pair<int, int>> checks = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
        int flip_count = 0;
        int m = mat.size(), n = mat[0].size();

        // start computing
        AllPossibles.push(mat);
        checked.insert(mat);
        while (!AllPossibles.empty()) {
            
            int CheckSize = AllPossibles.size();
            for (int i=0; i<CheckSize; i++) {
                
                if (CheckZero(AllPossibles.front())) return flip_count;
                else {

                    for (int row=0; row<m; row++) {
                        for (int col=0; col<n; col++) {

                            auto currMat = AllPossibles.front();
                            currMat[row][col] = 1 - currMat[row][col];

                            for (auto check: checks) {

                                int newRow = row + check.first;
                                int newCol = col + check.second;
                                if ((0 <= newRow) && (newRow < m) && (0 <= newCol) && (newCol < n))
                                    currMat[newRow][newCol] = 1 - currMat[newRow][newCol];

                            }
                            
                            if (checked.find(currMat) == checked.end()) {

                                AllPossibles.push(currMat);
                                checked.insert(currMat);

                            }

                        }
                    }
                    AllPossibles.pop();

                }

            }
            flip_count++;

        }

        return -1;

    }
};
